!!! FEATURE: Update auto-created ``Root.fusion`` to use prototypes instead of the ``page`` object

gerhard-boden opened the pull request:
A freshly created site package will now create a page protoype that extends `Neos.Neos:Page`. Using `/page` can be seen as deprecated and packages that rely on extending `page` should be updated to extend the page prototype instead. 

As a workaround the line `page = Neos.NodeTypes:Page` can be added to the `Root.fusion` as a fallback so extension points using the the `page` path as well as layout properties will still be in effect.

For further information see: http://neos.readthedocs.io/en/stable/HowTos/RenderingCustomDocumentTypes.html

kitsunet commented:
That's like super breaking and I am against it because it leaves third party packages without any default "add-on" point. Currently they can rely on the "page" path to exist.
The prototype is not a bad idea, but I would suggest that there is still "page" created from that prototype.

hlubek commented:
@kitsunet Wasn't the add-on point the PrimaryContent prototype for plugins wishing to contribute to the content? The "page" path cannot really be relied on because with a custom root matcher it would not be used.

The prototype `Page` is IMHO the best extension point for plugin packages to add content / metadata / resources to the rendered document. Because then it would be used for

* every root matcher rendering a `Page` (including the prototype based documents)
* all paths that instantiate a prototype extended from `Page`

kitsunet commented:
Right, fair enough. We could do it that way. It's still breaking though.

kitsunet commented:
I still feel it's bad to have no typoscript path to your page anymore, maybe it's just me and if everyone else says it's fine I will go with it...

gerhard-boden commented:
Thanks @kitsunet and @hlubek for your thoughts. When I submitted this PR already thought that it could be controversial :) So maybe we can add a `Discussion` flag to this.

In my personal experience using a prototype in my `roots.ts2` made it so much easier for me to create different page versions, while keeping the amount of repetitive code in TS2 files and templates to practically zero.

When I made my first steps in Neos, customizing the page object felt really inconvenient (maybe that also had to do with my very limited knowledge about TypoScript in general).

> That's like super breaking 
Since this change does not affect the existing `root.ts2`  within the site package, I don't feel like it's super breaking. :) You could update any page without any side effects. 

The only thing that happens is that the autogenerated `root.ts2` looks different when you kickstart a new site package. And even that being the case, you can still override it with a classic `page` object if you want to.

The reason why I proposed this PR in the first place is to spread the word about this (currently alternative) approach how do deal with the `Page` object, because it made things easier for me and I think it will also for many others who do not know about it yet.

dimaip commented:
Well, this PR works exactly as we have agreed on what the best practice for rendering web pages is, so there's no reason to not to do it. I think the demo site should be adjusted in the same way, as it currently promotes outdated patterns as well.

johannessteu commented on TYPO3.Neos.Kickstarter/Resources/Private/Generator/TypoScript/NodeTypes/Pages/Page.ts2:
```diff
@@ -0,0 +1,11 @@
+prototype(TYPO3.Neos.NodeTypes:Page.Document) < prototype(Neos:DefaultPage) {
```
This would add a new dependency to the `TYPO3.Neos.NodeTypes` Package, right? Not sure if we should add also a default for the default Page/Document or but this content into that prototype?

gerhard-boden replied:
Not sure if that's a big issue, since this kickstarter package is part of the Neos base distribution.
I guess 99% of all Neos projects utilize `TYPO3.Neos.NodeTypes:Page` in some way and therefore the `TYPO3.Neos.NodeTypes` package. The 1% that does not can still adapt that line to their own needs when the start a new project.

bwaidelich replied:
I think it is a big issue!
Much like the `TYPO3.Party` package I try to avoid the `NodeTypes` package where possible and I came across a couple of agencies and Neos devs that do the same (why do you think it's just 1%?)
It's a good place to get started but at least for a complex scenario it will never exactly fit and rewiring/disabling existing NodeTypes can get quite cumbersome and confusing

gerhard-boden replied:
The 99% part was just a guess. Guess i was wrong :smile: 

I meant not "a big issue", like in "it won't hurt anything". It's not like your site will break because you do not use the "NodeTypes" package with the above TypoScript definition in place. In my eyes it's just a little help to get people started. Like we already have here:
https://github.com/neos/neos-development-collection/blob/master/TYPO3.Neos.Kickstarter/Resources/Private/Generator/Configuration/NodeTypes.yaml#L8

We could also take out the lines and just leave them in as a comment.

bwaidelich replied:
Hehe, who knows maybe it's _me_ who's wrong ;)
Anyways, I'd like to avoid dependencies where possible - but as you said, it's not an issue in this case probably!

gerhard-boden replied:
>Anyways, I'd like to avoid dependencies where possible 
Totally agree with you! 

I guess kickstarting stuff is always tricky because we have to find a solution that works best for the majority of users out of the box.

aertmann replied:
the nodetypes package is already a dependency of the kickstarter which makes sense.. https://github.com/gerhard-boden/neos-development-collection/blob/master/TYPO3.Neos.Kickstarter/Resources/Private/Generator/Content/Sites.xml#L6 https://github.com/gerhard-boden/neos-development-collection/blob/master/TYPO3.Neos.Kickstarter/composer.json#L9

so no need to complicate this change with that.. if we want to change that, that's a whole different topic,,

dfeyer replied:
It's fine that the kickstarter as dependency to the NodeTypes package, but please try to keep Neos as decoupled from the NodeType package as possible (if not not completly)

We also start to use raw Neos without the NodeTypes package, and it's a pretty nice practice, that we can promote (or at least not block)

johannessteu commented on TYPO3.Neos.Kickstarter/Resources/Private/Generator/TypoScript/NodeTypes/Pages/Page.ts2:
```diff
@@ -0,0 +1,11 @@
+        content {
+            // The default content section
+            main = TYPO3.Neos:PrimaryContent {
+            nodePath = 'main'
```
Wrong/Missing indentation

gerhard-boden replied:
whoops, fixed!

johannessteu commented:
> I still feel it's bad to have no typoscript path to your page anymore, maybe it's just me and if everyone else says it's fine I will go with it...

I think it's ok to remove the static path for your page. But we need to make it somehow more visible. If i talk to some colleagues they would all say that page is the first entry point and like 95% are not aware of the root matcher and how the page object itself is generated or why the path `/page` is used. 

In general i do like this PR and the idea of it but i think it will break a lot of things because ppl really do rely atm on the `/page` path. So some kind of fallback would be great so this PR wouldn't break that much.

kitsunet commented:
Yep, I thought a bit more about it and also like it more and more :)

gerhard-boden commented:
> I think the demo site should be adjusted in the same way, as it currently promotes outdated patterns as well. 

@dimaip correct! That will be phase 2 :)

>In general i do like this PR and the idea of it but i think it will break a lot of things because ppl really do rely atm on the /page path. So some kind of fallback would be great so this PR wouldn't break that much.

@johannessteu IMHO the easiest way to a fallback is to rewrite your kickstarted site package to do things the old way in case one of your plugins relies on it. It's just a couple of lines, can be done in < 30 seconds. I could also add a comment about that in the `root.ts2`.

johannessteu commented:
> Yep, I thought a bit more about it and also like it more and more :)

So @kitsunet you are going to remove your 👎  ?

aertmann commented:
seems okay to me as it is better than before and an okay generic solution, although not how I'd do it personally

gerhard-boden commented:
Thanks for your review @aertmann! 
>although not how I'd do it personally 

always open for improvement, any suggestions I can implement?

dfeyer commented:
Fine for me thanks for the change @gerhard-boden

kitsunet commented on TYPO3.Neos.Kickstarter/Resources/Private/Generator/TypoScript/Root.ts2:
```diff
@@ -1,37 +1,41 @@
-		parts {
-			menu = TYPO3.Neos:Menu
-		}
+prototype(Neos:DefaultPage) < prototype(Page) {
```
What is `Neos` supposed to be? We generally always use packageKeys there so IMHO

`prototype({packageKey}:DefaultPage) < prototype(Page) {`

kitsunet replied:
What is `Neos` supposed to be? We generally always use packageKeys there so IMHO

`prototype({packageKey}:DefaultPage) < prototype(Page) {`

gerhard-boden replied:
Changed it according to your suggestion (also for the autogenerated `Page.ts2`)

The reason why I came up with `Neos:DefaultPage` was to have a generic name for the `DefaultPage` prototype.  So it would be easier to copy & paste stuff (especially when it comes to other node types that extend this prototype).

kitsunet commented on TYPO3.Neos.Kickstarter/Resources/Private/Generator/TypoScript/Root.ts2:
```diff
@@ -1,37 +1,41 @@
-	}
+root {
+    default {
+        type = ${q(node).property('_nodeType') + '.Document'}
```
Do I miss something or is that wrong right now and needs to be adapted to

`Neos:DefaultPage` (or rather `{packageKey}:DefaultPage`)

kitsunet replied:
This matcher should probably move to core in next major!?

gerhard-boden replied:
IMHO we could even wait until the next major after that, so people have more time to adapt. Because when we move this to the core, sites will actually start breaking after the update. Unless we can implement it in a way, that we somehow check how the `root.ts2` of the active site package is structured and only activate the matcher if it makes sense.

gerhard-boden commented:
Made some final adjustments, thanks for all the feedback guys! Really great to see so many having a look at this (8 including the bot smile:).

@aertmann @kitsunet  could you update your votes on the top with this last change (unless there is still something to do of course).

hlubek commented:
@kitsunet Do you mind to remove your -1?

gerhard-boden commented:
After a discussion with @mficzel at the sprint today I also agree that we should get the infrastructure ready for this change, before merging. So not relevant for Neos 3.0 anymore, since time is of essence to get more important things ready.

kitsunet commented:
Ok how do we proceed here? I agree by now that this is a good idea. Any plans on this @mficzel ?

gerhard-boden commented:
@kitsunet It's been a while, but if remember correctly the described Eel Helper and fusion Core change would be fairly easy to implement. I could take care. Would be put this change into 3.2? Since this is rather breaking I'd rather see it in 4.0?

hlubek commented:
I'm all in for changing this behavior in 4.0. I second that we need a way to check for the prototype existence and have a graceful fallback if that is not defined. We could amend \Neos\Fusion\FusionObjects\RendererImplementation to do that with a new property (e.g. `fallbackRenderer`), because Eel helpers don't have access to the runtime to check for `canRender` of the path for the type.

kitsunet commented:
@gerhard-boden could you update this soonish to latest master so we can get it finally in for this major?

gerhard-boden commented:
@kitsunet rebase is done! Will start working on implementing the ``fallbackRenderer``.

aertmann commented on Neos.SiteKickstarter/Classes/Service/GeneratorService.php:
```diff
@@ -106,8 +107,31 @@ protected function generateSitesFusion($packageKey, $siteName)
+
+        $fileContent = $this->renderTemplate($templatePathAndFilename, $contextVariables);
+
+        $sitesPageFusionPathAndFilename = $this->packageManager->getPackage($packageKey)->getResourcesPath() . 'Private/TypoScript/NodeTypes/Page.ts2';
```
still has TypoScript instead of Fusion

aertmann replied:
also .ts2 instead of .fusion

aertmann commented on Neos.SiteKickstarter/Resources/Private/Generator/Fusion/NodeTypes/Pages/Page.ts2:
```diff
@@ -0,0 +1,11 @@
+prototype(Neos.Neos:Document.Document) < prototype({packageKey}:DefaultPage) {
```
.ts2 instead of .fusion

aertmann commented on Neos.SiteKickstarter/Classes/Service/GeneratorService.php:
```diff
@@ -106,8 +107,31 @@ protected function generateSitesFusion($packageKey, $siteName)
+     */
+    protected function generateSitesPageFusion($packageKey, $siteName)
+    {
+        $templatePathAndFilename = 'resource://Neos.SiteKickstarter/Private/Generator/Fusion/NodeTypes/Pages/Page.ts2';
```
.ts2 instead of .fusion

gerhard-boden commented:
@aertmann Thank you for keeping an eye on this! 
I didn't even test it after the rebase, because flow command line output is broken on master
https://github.com/neos/flow-development-collection/pull/1195 

Maybe eagle eye Aske could have a look at that fix as well 😉

mficzel commented on Neos.SiteKickstarter/Resources/Private/Generator/Fusion/Root.fusion:
```diff
@@ -1,37 +1,39 @@
 /**
```
I think this file should be removed entirely and the default fusion should be adjusted to support document type based rendering.

```
root {
    ... 
    document {
		@position = 'end 9998'
		condition = Neos.Fusion:PrototypeExists {
		    prototypeName = ${q(node).property('_nodeType.name')}
		}
		type = ${q(documentNode).nodeType.name}
	}
	
	default {
		@position = 'end 9999'
		condition = TRUE
		renderPath = '/page'
	}
	...
}
```

Offcourse the `Neos.Fusion:PrototypeExists` object does not exist yet and this would also lead to conflicts with the default expectation there is a template for every  document type.

mficzel replied:
The breakiness could be reduced if we add extend the check to wether the prototyoe that is named after the document extends page 

```
root {
    ... 
    document {
        @position = 'end 9998'
        condition = Neos.Fusion:VerifyPrototype {
            name = ${q(node).property('_nodeType.name')}
            basedOn = `Neos.Neos:Page`
        }
        type = ${q(documentNode).nodeType.name}
    }
```

mficzel commented on Neos.SiteKickstarter/Resources/Private/Generator/Fusion/NodeTypes/Pages/Page.fusion:
```diff
@@ -0,0 +1,11 @@
+prototype(Neos.Neos:Document.Document) < prototype({packageKey}:DefaultPage) {
```
This looks somehow wrong, imho this should read:
`prototype({packageKey}:Page)) < prototype(Neos.Neos:Document) {`
If we do this we have to add the NodeType for that aswell.

mficzel replied:
Also i would like to create to a `Site` and a `Page` NodeType by default, with constraints that allow only page and shortcut inside root.  That would require two fusion prototypes aswell.

gerhard-boden commented:
Rebased and did some minor clean up and fixing in preparation for the major release. This PR is now stripped down to it's essentials (also removed the root matcher from the autogenerated file, since we will take care of that within the core). 

@mficzel How about monday for our `Neos.Fusion:PrototypeExists` meeting? (let's coordinate in slack again next week)

gerhard-boden commented:
Just made some final adjustments, this looks very ready to me now. Note that this will only work as soon as all the necessary changes to the core are ready (hence why I keep the WIP label on it)
Since it's just a generator script though, there would be no immediate damage to master.

kitsunet commented:
@mficzel can you recheck this one?

gerhard-boden commented:
just one sec, this need a final touch up

gerhard-boden commented:
Final update for latest root matcher changes and still found some bugs (I was asking myself "How often can you get this simple thing wrong?!" 😆)

gerhard-boden commented on Neos.SiteKickstarter/Resources/Private/Generator/Fusion/Root.fusion:
```diff
@@ -1,37 +1,32 @@
-			sectionName = 'headScripts'
-		}
-	}
+prototype({packageKey}:DefaultPage) < prototype(Page) {
```
Thanks for the review, I think you saw this line in an outdated version. This has been updated 28 days ago https://github.com/neos/neos-development-collection/pull/615/commits/a5e7c9b946537d7301a5c19d0e6b0197b4218f80

kitsunet commented:
The above is not super cool. Also the PR message should contain information WHY and HOW this is breaking and how to mitigate it.

beheist commented on Neos.SiteKickstarter/Resources/Private/Generator/Configuration/NodeTypes.Document.Page.yaml:
```diff
@@ -2,7 +2,16 @@
+  superTypes:
+    'Neos.Neos:Document': true
+  ui:
+    icon: 'icon-file'
```
Add `label: "{packageKey} Custom Page"` here

beheist commented on Neos.SiteKickstarter/Resources/Private/Generator/Configuration/NodeTypes.Document.Page.yaml:
```diff
@@ -2,7 +2,16 @@
 #    'sidebar':
-#      type: 'Neos.Neos:ContentCollection'
\ No newline at end of file
+#      type: 'Neos.Neos:ContentCollection'
```
I'd move this comment down under the actual node type, so that when the user uncomments the markup is actually valid. Like this:

```
'Foo.Bar:Page':
  superTypes:
    'Neos.Neos:Document': true
  ui:
    label: "Foo.Bar Custom Page"
    icon: 'icon-file'
  childNodes:
    main:
      type: 'Neos.Neos:ContentCollection'
      
#   If you need more than just the "main" content area you can uncomment the snippet below. It will add a "sidebar" area
#   to your pages. You can add as many areas as you like, still only those that are rendered in your Fusion-code and
#   Template are visible on your page.

#    sidebar:
#      type: 'Neos.Neos:ContentCollection'
```

gerhard-boden commented:
@beheist Thanks for the review, adjusted the label to reduce confusion after kickstarting a package
